keps/2349-multikueue-external-custom-job-support/README.md

```diff
@@ -0,0 +1,186 @@
+Specified controller is used by workload, cluster and admission check MultiKueue reconcilers.
+
+### User Stories
+
+#### Story 1
+
+I would like to support the extension mechanism so that we can implement the MultiKueue controller for the custom / in-house Jobs.
+
+### Constraints
+
+Internally the MultiKueue uses a set of controller-runtime client, one for each MultiKueueCluster, the clients are used both by the MultiKueue internals and passed to the job adapters for the custom jobs synchronization.
+
+The clients cannot be share with an external controller and assuming the external controller can get the same connectivity to the worker clusters might be wrong, especially when we are taking about FS mounted kubeconfigs.
+
+Designing additional APIs to communicate between the internal and external controller will bring extra complexity and latency at runtime. 
+
+### Risks and Mitigations
+
+## Design Details
+
+Provide a way to integrate external controllers to manage Jobs in the Kueue.
+MulikueueCluster API needs to be modified to ensure that one multikueue cluster is only managed by one admission check controller.
+
+1. Add ControllerName to the `MultiKueueClusterSpec`
+```golang
+type MultiKueueClusterSpec struct {
+  ...
+	// controllerName is name of the controller which will actually perform
+	// the checks. This is the name with which controller identifies with,
+	// not necessarily a K8S Pod or Deployment name. Cannot be empty.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default="kueue.x-k8s.io/multikueue"
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="field is immutable"
+	ControllerName string `json:"controllerName"`
+  ...
+```
+
+2. Make Controller Name used in Multikueue admission check configurable.
+Extend MultiKueue Admission Check controller options to include controllerName.
+
+```golang
+type SetupOptions struct {
+  ...
+	controllerName    string
+  ...
+}
+
+// WithControllerName - sets the controller name for which the multikueue
+// admission check match.
+func WithControllerName(controllerName string) SetupOption {
+	return func(o *SetupOptions) {
+		o.controllerName = controllerName
+	}
+}
+
+func SetupControllers(mgr ctrl.Manager, namespace string, opts ...SetupOption) error {
+...
+}
+```
+
+3. Make the Adapters list for the MultiKueue Admission Check controller configurable.
+
+```golang
+type SetupOptions struct {
+  ...
+	adapters          map[string]jobframework.MultiKueueAdapter
+  ...
+}
+
+// WithAdapters - sets all the MultiKueue adapters.
+func WithAdapters(adapters map[string]jobframework.MultiKueueAdapter) SetupOption {
+	return func(o *SetupOptions) {
+		o.adapters = adapters
+	}
+}
+
+// WithAdapter - sets or updates the adapter of the MultiKueue adapters.
+func WithAdapter(adapter jobframework.MultiKueueAdapter) SetupOption {
+	return func(o *SetupOptions) {
+		o.adapters[adapter.GVK().String()] = adapter
+	}
+}
+```
+
```

Comment:
Please expand on:
- which subcomponents of MK each external controller is implementing (even if provided by a framework)
- how does a configuration for multiple of these controllers look like from the ClusterQueue point of view.